DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs

```diff
@@ -0,0 +1,294 @@
+﻿using DMCompiler.Compiler.DM;
+using DMCompiler.DM;
+using OpenDreamShared.Compiler;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+#nullable enable
+namespace DMCompiler.Compiler.DMPreprocessor {
+    /*
+     * NOTE: Some operations are (as in BYOND) not implemented for preproc expressions.
+     * This includes ternary, bitwise, and assignment.
+    Grammar: (As usual in extended BNF, {A} means 0 or more As, and [A] means an optional A. Brace characters are escaped with single-quotes.)
+    Expression ::= AndExp ['||' AndExp]
+    AndExp ::= CompExp ['&&' CompExp]
+    CompExp ::= LtGtExp [('==' | '!=') LtGtExp]
+    LtGtExp ::= AddSubExp [(‘<’ | ‘<=’ | ‘>’ | ‘>=’ | ‘==’ | ‘!=’) AddSubExp]
+    AddSubExp ::= MultExp [('+' | '-') MultExp]
+    MultExp ::= PowerExp [('*' | '/' | '%') PowerExp]
+    PowerExp ::= UnaryExp [** PowerExp]
+    UnaryExp ::= {!}SignExp
+    SignExp ::= [-]Primary
+    Primary ::= '(' Expression ')' | DefinedExp | Constant
+    DefinedExp ::= 'defined(' MacroIdentifier ')'
+    Constant ::= Integer | Float | Null
+    */
+
+    /// <summary>
+    /// An extremely simple parser that acts on Preproc tokens, <br/>
+    /// held separate from DMParser because of slightly different behaviour and far simpler implementation.
+    /// </summary>
+    internal static class DMPreprocessorParser {
```

Comment:
Is there a reason this can't extend the `Parser<T>` class our other parsers use?

Reply:
Most practically, I want to make DMPreprocessorParser static eventually and a static object cannot inherit from a non-static (this would be the only stateless parser in the codebase)

Second mostly practically, DMPP doesn't really make use of any of the code available in DMParser or Parser<>, methods of the same name would need to be completely overridden, defeating the purpose of inheritance as a way of reducing code size & complexity.

If it's annoying that it's the odd one out, it could make sense to call it the DMPreprocessorEvaluator instead or something, since it is doing parsing and evaluation in one stroke.